### PR TITLE
[faiss v1.7.1] Impl PreTransformIndex

### PR DESCRIPTION
**Pull request by ava57r**

### CODE REVIEW

**Comment by ava57r** [comment]
https://github.com/facebookresearch/faiss/pull/1816

and 
https://github.com/facebookresearch/faiss/pull/1869

merged

File: src/index/pretransform.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            inner: inner_ptr as *mut FaissIndexPreTransform,
+        }
+    }
+}
+
+impl_native_index!(PreTransformIndexImpl);
+
+impl_native_index_clone!(PreTransformIndexImpl);
+
+impl ConcurrentIndex for PreTransformIndexImpl {
```
Comment (Enet4) [inline src/index/pretransform.rs:100]: This one should only be implemented if the sub-index is also a concurrent index. We can keep track of the underlying index type `I` through phantom data. The complete version would be:

```rust
struct PreTransformIndexImpl<'a, I> {
    inner: *mut FaissIndexPreTransform,
    phantom: std::marker::PhantomData<(&'a (), I)>,
}
```

Then `into_pre_transform(self)` would return a `PreTransformIndexImpl<'static, Self>`.

**Review by Enet4 — CHANGES_REQUESTED** [review]
While most of the constructs were well thought out, but this `PreTransform` can be exploited to perform memory unsafety.

- Values can outlive its native objects if they are not owned.
- `ConcurrentIndex` is implemented even if the native index does not.

More details and guidance in the comments below.

**Comment by ava57r** [comment]
Maybe we should not use the option when `own_fields = = false`?

**Comment by Enet4** [comment]
> Maybe we should not use the option when `own_fields = = false`?

I mean, I suppose part of the problem would no longer apply if we require all indexes to be owned (and so `own_fields` would never be `false`), but adding a few type/lifetime parameters should do the trick here.

**Comment by ava57r** [comment]
I won't complicate the impl. I like the transfer of ownership.

**Comment by ava57r** [comment]
I suggest starting with the ownership of indexes, quantizers. In the future, it will become clear whether to use `own_fields = = false`

**Review by Enet4 — APPROVED** [review]
Thank you.

**Pull request merged**